packages/graphql/src/services/sessions.ts

```diff
@@ -1,13 +1,132 @@
-import { SessionStatus } from '@klicker-uzh/prisma'
+import {
+  Question,
+  QuestionType,
+  SessionBlockStatus,
+  SessionStatus,
+} from '@klicker-uzh/prisma'
 import { ContextWithUser } from '@lib/context'
-import { getRedis } from '../lib/redis'
 
-export async function startSession(
-  { id }: { id: string },
+function processQuestionData(question: Question): AllQuestionTypeData {
+  switch (question.type) {
+    case QuestionType.SC:
+    case QuestionType.MC: {
+      return {
+        ...question,
+        options: question.options!.valueOf(),
+      } as ChoicesQuestionData
+    }
+
+    default:
+      return question
+  }
+}
+
+function prepareInitialInstanceResults(questionData: AllQuestionTypeData) {
+  switch (questionData.type) {
+    case QuestionType.SC:
+    case QuestionType.MC: {
+      const choices = questionData.options.choices.reduce(
+        (acc, _, ix) => ({ ...acc, [ix]: 0 }),
+        {}
+      )
+      return { choices }
+    }
+
+    default:
+      return {}
+  }
+}
+
+interface BlockArgs {
+  questionIds: number[]
+  randomSelection?: number
+  timeLimit?: number
+}
+
+interface CreateSessionArgs {
+  name: string
+  displayName?: string | null
+  blocks: BlockArgs[]
+}
+
+export async function createSession(
+  { name, displayName, blocks }: CreateSessionArgs,
   ctx: ContextWithUser
 ) {
-  const redis = getRedis()
+  const allQuestionsIds = blocks.reduce<number[]>(
+    (acc, block) => [...acc, ...block.questionIds],
+    []
+  )
+
+  const questions = await ctx.prisma.question.findMany({
+    where: {
+      id: {
+        in: allQuestionsIds,
+      },
+    },
+    include: {
+      attachments: true,
+    },
+  })
+
+  const data = {
```

Comment:
Looks a bit complex, but is, in principle, just a db "transaction". Creates a session alongside all of the required session blocks and question instances inside of these blocks. Does it all at once, or not at all. And will automatically create the relationships from "Session - SessionBlock - QuestionInstance".